### PR TITLE
Only apply network_fail_open policy to 5xx Http status code

### PR DESCRIPTION
**Pull request by qiwzhang**
Signed-off-by: Wayne Zhang <qiwzhang@google.com>

If ServiceControl  error response has body, it may contain important debug info, also pass that info to the client.

### CODE REVIEW

File: src/api_manager/service_control/aggregated.cc [hunk 1]
```diff
@@ -371,8 +371,9 @@ void Aggregated::Check(
           *response, service_control_proto_.service_name(), &response_info);
       on_done(status, response_info);
     } else {
-      // If network_fail_open is true, it is always OK to proceed.
-      if (network_fail_open_) {
+      // Only Code::UNAVAILABLE error is network failure.
+      // If network_fail_open is true, it is OK to proceed
+      if (network_fail_open_ && status.error_code() == Code::UNAVAILABLE) {
         on_done(Status::OK, response_info);
```
Comment (nareddyt) [inline src/api_manager/service_control/aggregated.cc]: Consider logging a warning/error about network fail open

File: src/api_manager/service_control/aggregated.cc [hunk 2]
```diff
@@ -371,8 +371,9 @@ void Aggregated::Check(
           *response, service_control_proto_.service_name(), &response_info);
       on_done(status, response_info);
     } else {
-      // If network_fail_open is true, it is always OK to proceed.
-      if (network_fail_open_) {
+      // Only Code::UNAVAILABLE error is network failure.
```
Comment (nareddyt) [inline src/api_manager/service_control/aggregated.cc]: Can you clarify where `Unavailable` is set for non 4xx responses? And what is set for 4xx responses?

Is it here? https://github.com/cloudendpoints/esp/blob/28c924749263c0a8fe252ed3f2c678cb3988534a/src/nginx/http.cc#L993

File: src/api_manager/service_control/aggregated.cc [hunk 3]
```diff
@@ -371,8 +371,9 @@ void Aggregated::Check(
...
+      // If network_fail_open is true, it is OK to proceed
+      if (network_fail_open_ && status.error_code() == Code::UNAVAILABLE) {
```
Comment (TAOXUY) [inline src/api_manager/service_control/aggregated.cc]: Would be better to limit to all [5XX](https://www.websitepulse.com/kb/5xx_http_status_codes) errors not only service unavaliable? 
Also, please change the title to be
"Make fail open only on server side problem"

File: src/api_manager/service_control/aggregated.cc (hunk 2 above)
Comment (qiwzhang) [reply src/api_manager/service_control/aggregated.cc]: 1) [here](https://github.com/cloudendpoints/esp/blob/master/src/nginx/http.cc#L573)  Http status code is copied to Status()

2) [here](https://github.com/cloudendpoints/esp/blob/master/src/api_manager/service_control/aggregated.cc#L607), we convert http status to RpcStatus::code in ToProto() call.  This code is also modifed in this PR.

File: src/api_manager/service_control/aggregated.cc (hunk 1 above)
Comment (qiwzhang) [reply src/api_manager/service_control/aggregated.cc]: Done

File: src/api_manager/service_control/aggregated.cc (hunk 3 above)
Comment (qiwzhang) [reply src/api_manager/service_control/aggregated.cc]: Done

**Comment by qiwzhang** [comment]
PTAL

File: src/api_manager/service_control/aggregated.cc [hunk 4]
```diff
@@ -93,6 +93,16 @@ const char servicecontrol_service[] =
 const char quotacontrol_service[] =
     "/google.api.servicecontrol.v1.QuotaController";
 
+// Define network failure error codes:
+// All 500 Http status codes are marked as network failure.
```
Comment (TAOXUY) [inline src/api_manager/service_control/aggregated.cc]: 500 causes ambiguity of `Internal Server Error`. Please use 5XX

Comment (qiwzhang) [reply src/api_manager/service_control/aggregated.cc]: Done

**Pull request merged**